ibis/expr/types/arrays.py

```diff
@@ -383,7 +385,45 @@ def join(self, sep: str | ir.StringValue) -> ir.StringValue:
         """
         return ops.ArrayStringJoin(self, sep=sep).to_expr()
 
-    def map(self, func: Deferred | Callable[[ir.Value], ir.Value]) -> ir.ArrayValue:
+    def _construct_array_func_inputs(self, func):
+        if isinstance(func, Deferred):
+            name = "_"
+            index = None
+            resolve = func.resolve
+        elif callable(func):
+            names = (
+                key
+                for key, value in inspect.signature(func).parameters.items()
+                # arg is already bound
+                if value.default is EMPTY
+            )
+            name = next(names)
+            index = next(names, None)
+            resolve = func
+        else:
+            raise TypeError(
+                f"function must be a Deferred or Callable, got `{type(func).__name__}`"
+            )
+
+        shape = self.op().shape
+        parameter = ops.Argument(name=name, shape=shape, dtype=self.type().value_type)
+
+        kwargs = {name: parameter.to_expr()}
+
+        if index is not None:
+            index_arg = ops.Argument(name=index, shape=shape, dtype=dt.int64)
+            kwargs[index] = index_arg.to_expr()
+            index = index_arg.param
+
+        body = resolve(**kwargs)
+        return parameter.param, index, body
+
+    def map(
+        self,
+        func: Deferred
+        | Callable[[ir.Value], ir.Value]
+        | Callable[[ir.Value, ir.Value], ir.Value],
+    ) -> ir.ArrayValue:
```

Comment:
Can't comment on the direct line - could you add a bit about the possible signatures of `func` below in the `Parameters` section? May also want to include that the index is 0-based.

Same goes for `filter` below.

Reply:
Done!